sdk/src/signer/mod.rs

```diff
@@ -103,6 +109,25 @@ pub fn unique_signers(signers: Vec<&dyn Signer>) -> Vec<&dyn Signer> {
     signers.into_iter().unique_by(|s| s.pubkey()).collect()
 }
 
+/// The `EncodableKey` trait defines the interface by which cryptographic keys/keypairs are read,
+/// written, and derived from sources.
+pub trait EncodableKey: Sized {
+    fn pubkey_string(&self) -> Result<String, Box<dyn error::Error>>;
+    fn read_key<R: Read>(reader: &mut R) -> Result<Self, Box<dyn error::Error>>;
+    fn read_key_file<F: AsRef<Path>>(path: F) -> Result<Self, Box<dyn error::Error>>;
+    fn write_key<W: Write>(&self, writer: &mut W) -> Result<String, Box<dyn error::Error>>;
+    fn write_key_file<F: AsRef<Path>>(&self, outfile: F) -> Result<String, Box<dyn error::Error>>;
+    fn key_from_seed(seed: &[u8]) -> Result<Self, Box<dyn error::Error>>;
+    fn key_from_seed_and_derivation_path(
+        seed: &[u8],
+        derivation_path: Option<DerivationPath>,
+    ) -> Result<Self, Box<dyn error::Error>>;
+    fn key_from_seed_phrase_and_passphrase(
+        seed_phrase: &str,
+        passphrase: &str,
+    ) -> Result<Self, Box<dyn error::Error>>;
```

Comment:
Will these all be relevant for ElGamal keypairs and authenticated encryption keys?

Reply:
Yes these functions are all relevant though for `key_from_seed_and_derivation_path`, I was thinking of outputting a suitable error type like [this](https://github.com/solana-labs/solana/pull/30901/files#diff-7d059813b85f149efcc5236cf7f8c29645b18bb9787cdaa05cec84711b06a8a6R309). It is not that `key_from_seed_and_derivation_path` does not make sense for ElGamal keypairs or authenticated encryption keys, but I am not sure how useful it would be and there is also no crate like `ed25519-dalek-bip32` that can do the parsing for us.